File: cpp/chapter_stack_and_queue/linkedlist_deque.cpp

#include "../utils/common.hpp"

/* 双向链表节点 */
struct DoublyListNode {
    int val;
    DoublyListNode *prev;   // 前驱指针
    DoublyListNode *next;   // 后继指针
    DoublyListNode(int x) : val(x), prev(nullptr), next(nullptr) {}
};

/* 基于双向链表实现的双向队列 */
class LinkedListDeque {
  private:
    DoublyListNode *front;   // 队首指针
    DoublyListNode *rear;   // 队尾指针
    int queSize=0;            // 队列的大小
  public:
    // 构造函数
    LinkedListDeque():front(nullptr),rear(nullptr){}
    // 析构函数 
    ~LinkedListDeque() {
        DoublyListNode *p = front;
        while (p != nullptr) {
            DoublyListNode *tmp = p;
            p = p->next;
            delete tmp;
        }
    }
    // 获取队列的大小
    int size() {
        return queSize;
    }
    // 判断队列是否为空
    bool isEmpty() {
        return queSize == 0;
    }
    // 入队
    void push(int num,bool isFront){
        DoublyListNode *node = new DoublyListNode(num);
        if (isEmpty()) {
            front = node;
            rear = node;
        } else if (isFront) {
            node->next = front;
            front->prev = node;
            front = node;
        } else {
            node->prev = rear;
            rear->next = node;
            rear = node;
        }
        queSize++;
    }
    // 队首入队
    void pushFirst(int num) {
        push(num,true);
    }
    // 队尾入队
    void pushLast(int num) {
        push(num,false);
    }
    // 出队
    int pop(bool isFront) {
        if (isEmpty()) {
            return -1;
        }
        int val;
        if (isFront) {
            val = front->val;
            DoublyListNode *tmp = front;
            front = front->next;
            if (front != nullptr) {
                front->prev = nullptr;
            }
            delete tmp;
        } else {
            val = rear->val;
            DoublyListNode *tmp = rear;
            rear = rear->prev;
            if (rear != nullptr) {
                rear->next = nullptr;
            }
            delete tmp;
        }
        queSize--;
        return val;
    }
    // 队首出队
    int popFirst() {
        return pop(true);
    }
    // 队尾出队
    int popLast() {
        return pop(false);
    }
    // 获取队首元素
    int peekFirst() {
        if (isEmpty()) {
            return -1;
        }
        return front->val;
    }

    // 获取队尾元素
    int peekLast() {
        if (isEmpty()) {
            return -1;
        }
        return rear->val;
    }
    // 将队列转换为数组
    vector<int> toVector() {
        vector<int> res;
        DoublyListNode *p = front;
        while (p != nullptr) {
            res.push_back(p->val);
            p = p->next;
        }
        return res;
    }
    
};





int main() {
    /* 初始化双向队列 */
    LinkedListDeque *deque = new LinkedListDeque();
    deque->pushLast(3);
    deque->pushLast(2);
    deque->pushLast(5);
    cout << "双向队列 deque = ";
    printVector(deque->toVector());

    /* 访问元素 */
    int peekFirst = deque->peekFirst();
    cout << "队首元素 peekFirst = " << peekFirst << endl;
    int peekLast = deque->peekLast();
    cout << "队尾元素 peekLast = " << peekLast << endl;

    /* 元素入队 */
    deque->pushLast(4);
    cout << "元素 4 队尾入队后 deque =";
    printVector(deque->toVector());
    deque->pushFirst(1);
    cout << "元素 1 队首入队后 deque = ";
    printVector(deque->toVector());

    /* 元素出队 */
    int popLast = deque->popLast();
    cout << "队尾出队元素 = " << popLast << "，队尾出队后 deque = ";
    printVector(deque->toVector());
    int popFirst = deque->popFirst();
    cout << "队首出队元素 = " << popFirst << "，队首出队后 deque = ";
    printVector(deque->toVector());

    /* 获取双向队列的长度 */
    int size = deque->size();
    cout << "双向队列长度 size = " << size << endl;

    /* 判断双向队列是否为空 */
    bool isEmpty = deque->isEmpty();
    cout << "双向队列是否为空 = " << boolalpha << isEmpty << endl;

    // 释放内存
    delete deque;

    return 0;
}

File: python/chapter_stack_and_queue/linkedlist_deque.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
        self.prev = None
class LinkedListDeque:
    #基于链表实现的双端队列
    def __init__(self) -> None:
        self._front = None
        self._rear = None
        self._size = 0
    #长度
    def size(self)->int:
        return self._size
    #判断是否为空
    def is_empty(self)->bool:
        return self._size==0
    #入队
    def push(self,num:int,is_front:bool):
        node = ListNode(num)
        if self.is_empty():
            self._front = node
            self._rear = node
        elif is_front:
            node.next = self._front
            self._front.prev = node
            self._front = node
        else:
            node.prev = self._rear
            self._rear.next = node
            self._rear = node
        self._size+=1
    #出队
    def pop(self,is_front:bool)->int:
        if self.is_empty():
            raise Exception("Deque is empty")
        if is_front:
            val = self._front.val
            self._front = self._front.next
            if self._front:
                self._front.prev = None
            else:
                self._rear = None
        else:
            val = self._rear.val
            self._rear = self._rear.prev
            if self._rear:
                self._rear.next = None
            else:
                self._front = None
        self._size-=1
        return val
    #获取队头元素
    def pop_firts(self)->int:
        return self.pop(True)
    def pop_last(self)->int:
        return self.pop(False)
    #打印
    def to_list(self)->list[int]:
        res = []
        cur = self._front
        while cur:
            res.append(cur.val)
            cur = cur.next
        return res
if __name__ == "__main__":
    deque = LinkedListDeque()
    deque.push(1,True)
    deque.push(2,True)
    deque.push(3,True)
    print(deque.to_list())
    deque.push(0,False)
    print(deque.to_list())
    front:int = deque.pop_firts()
    print(front)
    rear:int = deque.pop_last()
    print(rear)
    print(deque.to_list())
    size:int = deque.size()
    print(size)
    is_empty:bool = deque.is_empty()
    print(is_empty) 